### PR TITLE
Semgrep/Azure AD SAML SSO guide

### PR DESCRIPTION
**Pull request by holdenmcg**
Hi team, 

this PR is for a guide for setting up Azure SSO with Active Directory under the 'SSO' section of the docs.

if you could run the docs and look through the guide as you review it, that would be splendid 🙏 tysm!!

image

# Thanks for improving Semgrep Docs 😀

### Please ensure

- [ ] A subject matter expert (SME) reviews the content
- [x] A technical writer reviews the content or PR
- [x] This change has no security implications or else you have pinged the security team
- [x] Redirects are added if the PR changes page URLs

### CODE REVIEW

File: docs/semgrep-app/sso.md [hunk 1]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
 
 If you encounter issues during the setup process, reach out to [support@r2c.dev](mailto:support@r2c.dev) for assistance.
 
-<MoreHelp />
+### Setting up SAML SSO with Azure Active Directory
+
+Set up Semgrep as an Enterprise App with SAML SSO using Azure's Active Directory service by following these steps:
+
+1. Enable **Enterprise App** in **Azure Active Directory** with SAML SSO selected. Follow the instructions in [Enable single sign-on for an enterprise application](https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/add-application-portal-setup-sso).
+2. Go to Semgrep App [Settings](https://semgrep.dev/orgs/-/settings/sso) page. TODO Under **Basic Configuration**, you'll need to enter the `Entity ID` and the `Reply URL`. These are provided in the Semgrep App on the left-side **Settings** tab, under the **SSO** tab on the top, as `Audience URL` and `SSO URL`, respectively. Make sure you switch the provider type to **SAML2** in the SSO form in the Semgrep App.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:69]: I don't see **Basic Configuration** in the Semgrep App. Do we mean **Basic SAML Configuration** on the side of Azure? We need to specify at the beginning of each step where we are. These steps are too long, there should be one or two actions at maximum for a user to do in each step.

File: docs/semgrep-app/sso.md [hunk 2]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
 
 If you encounter issues during the setup process, reach out to [support@r2c.dev](mailto:support@r2c.dev) for assistance.
 
-<MoreHelp />
+### Setting up SAML SSO with Azure Active Directory
+
+Set up Semgrep as an Enterprise App with SAML SSO using Azure's Active Directory service by following these steps:
+
+1. Enable **Enterprise App** in **Azure Active Directory** with SAML SSO selected. Follow the instructions in [Enable single sign-on for an enterprise application](https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/add-application-portal-setup-sso).
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:68]: Which section or sections of the linked document should we read to find this information?

File: docs/semgrep-app/sso.md [hunk 3]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
...
+2. Go to Semgrep App [Settings](https://semgrep.dev/orgs/-/settings/sso) page. TODO Under **Basic Configuration**, you'll need to enter the `Entity ID` and the `Reply URL`. These are provided in the Semgrep App on the left-side **Settings** tab, under the **SSO** tab on the top, as `Audience URL` and `SSO URL`, respectively. Make sure you switch the provider type to **SAML2** in the SSO form in the Semgrep App.
+    !["Entity ID" and "Reply URL" fields in Semgrep App](/img/azure-sso-fields-2.png ""Entity ID" and "Reploy URL" fields from within the App")
+3. TODO You need to add two attribute statements called `name` and `email`, making sure to set the `email` attribute to `user.mail`. For help regarding **Attributes and Claims**, refer to [Customize claims issued in the SAML token for enterprise applications](https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-saml-claims-customization) documentation.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:71]: Great that we are linking to specific docs. Which section or sections of the linked document should we read to find this information?

File: docs/semgrep-app/sso.md [hunk 4]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
...
+4. In the Azure Single sign-on page, find the `Login URL`, `Azure AD Identifier`, and `Certificate (Base64)`. Pate them to your [Settings](https://semgrep.dev/orgs/-/settings/sso) page of SAML2 SSO as `IdP SSO URL` (`Login URL`), `IdP Issuer ID` (`Azure AD Identifier`), and `X.509 Certificate` (`Certificate (Base64)`).
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:72]:
```suggestion
4. In the Azure Single sign-on page, find the `Login URL`, `Azure AD Identifier`, and `Certificate (Base64)`. Paste them to your [Settings](https://semgrep.dev/orgs/-/settings/sso) page of SAML2 SSO as `IdP SSO URL` (`Login URL`), `IdP Issuer ID` (`Azure AD Identifier`), and `X.509 Certificate` (`Certificate (Base64)`). 
```

File: docs/semgrep-app/sso.md [hunk 5]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
...
+    !["Entity ID" and "Reply URL" fields in Semgrep App](/img/azure-sso-fields-2.png ""Entity ID" and "Reploy URL" fields from within the App")
+3. TODO You need to add two attribute statements called `name` and `email`, making sure to set the `email` attribute to `user.mail`. For help regarding **Attributes and Claims**, refer to [Customize claims issued in the SAML token for enterprise applications](https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-saml-claims-customization) documentation.
+4. In the Azure Single sign-on page, find the `Login URL`, `Azure AD Identifier`, and `Certificate (Base64)`. Pate them to your [Settings](https://semgrep.dev/orgs/-/settings/sso) page of SAML2 SSO as `IdP SSO URL` (`Login URL`), `IdP Issuer ID` (`Azure AD Identifier`), and `X.509 Certificate` (`Certificate (Base64)`). 
+    ![Setting additional SSO form fields](/img/azure-sso-app-fields.png "Setting additional SSO form fields")
+5. TODO At the top of the form, fill in your email domain and a display name for the SSO integration.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:74]: This seems to be prefilled already (I checked in two organizations).

File: docs/semgrep-app/sso.md [hunk 6]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
...
+    ![Display name and email domain in the SSO form](/img/azure-sso-app-fields-3.png "Display name and email domain in the SSO form")
+6. Once all the fields are filled out, press the **Create/Update Auth Provider** button at the bottom of the form. See a green confirmation message in the upper-left of the app to verify that we've recieved your SSO info.
+    ![Submitting your SSO info in the App](/img/azure-sso-submit.png "Submitting your SSO info in the App")
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:77]: We do not have to include this screenshot...
```suggestion
```

File: docs/semgrep-app/sso.md [hunk 7]
```diff
@@ -61,4 +61,21 @@ To set up SSO:
...
+
+To add users to the application in so they can log in with their domain emails: refer to the [Assign users and groups to an application](https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/assign-user-or-group-access-portal).
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:79]: So this change is only done on the side of Azure? Our users do not have to enable anything in Semgrep App?

**Comment by s-santillan** [comment]
Just an update - I asked permission from @holdenmcg to carry this update into the finish line and will be replying to / addressing @adamkvitek 's comments hopefully after finishing up 1 other task. Cheers yall. :)

File: docs/semgrep-app/sso.md (hunk 2 above)
Comment (s-santillan) [reply docs/semgrep-app/sso.md]: outdated - resolving as "addressed in subsequent commits by me"

File: docs/semgrep-app/sso.md (hunk 3 above)
Comment (s-santillan) [reply docs/semgrep-app/sso.md]: resolving as "addressed in subsequent commits by me"

File: docs/semgrep-app/sso.md (hunk 5 above)
Comment (s-santillan) [reply docs/semgrep-app/sso.md]: FYI, on my end it was not prefilled @adamkvitek . Probably depends on the version of Semgrep App when the org was added.

**Comment by s-santillan** [comment]
Notes on some patterns in this doc:

* I mostly used Holden's steps after observing that Azure AD's docs did not offer a golden path from start to finish. In fact, the top official docs link when googling "Create a custom enterprise app with Azure AD" is to **add already existing apps**.
* Certain steps require screenshots as checkpoints so the user knows they are on the right track, so I kept some of Holden's.

WRT my changes from Holden's:
* Sometimes the steps say fill out Y and Z with values from A and B **respectively**. This is easy to misread so I separated them into steps, such as: Copy value from A and paste into Y.
* I cut up the steps into 3 general procedures with headers so that the users could follow along easier.
* I checked some docs and noticed that the **UI for Azure AD has changed a few times so I removed all descriptions to the graphical locations /layout of the menu**. Instead I used the IA of the page, such as Headers.  For example, "Under Manage, click SomeItem."
* Yes we use screenshots but even if they change users can follow along by checking the location of something through the page IA, and it's easier for us to catch and keep up-to-date.

**Comment by s-santillan** [comment]
@adamkvitek This is ready for review 🙏 

1. Should we make it into its own document? It feels like it should be its own document...
2. I'm not sure about my choice of words for the copy/paste procedure (Setting up SAML SSO for your new Enterprise App).
3. I did try to reduce some screenshots but included those that the user is likely to need as a "checkpoint".

File: docs/semgrep-app/sso.md [hunk 8]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+
+You have now created a custom Enterprise App for Semgrep to integrate into Azure AD. This enables you to set up SAML SSO.
+
+#### Setting up SAML SSO for your new Enterprise App
+
+1. From your new Enterprise App's page, Click **Single-sign on** > **SAML** tile. This takes you to the **Set up Single Sign-On with SAML** page.
+![Screenshot of your Enterprise App's page](/img/azure-select-sso.png#bordered)
+2. Under the **Basic SAML Configuration** form, click **Edit** to enter values for **Entity ID** and **Reply URL**. These values are retrieved from Semgrep App by performing the following steps:
+    1. Go to Semgrep App [Settings](https://semgrep.dev/orgs/-/settings/sso) page, and then click **SAML2** under **SSO Method** to display the SAML2 form.
+    2. Copy the **Audience URL** value from Semgrep App to **Entity ID**.
```
[the remark this answers is not in the document]
Comment (adamkvitek) [reply docs/semgrep-app/sso.md:106]: @s-santillan feel free to simplify my suggestion, I tried to do my best... but it is a complex UI.

File: docs/semgrep-app/sso.md [hunk 9]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+    3. Copy the **SSO URL** value from Semgrep App to **Reply URL**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:107]:
```suggestion
    3. Copy the **SSO URL** value from Semgrep App, then in **Basic SAML Configuration**, under **Reply URL (Assertion Consumer Service URL)** click the **Add reply URL** and paste in the URL.
```

File: docs/semgrep-app/sso.md [hunk 10]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+3. In the **Basic SAML Configuration** form, click **Save**.
+4. Under the **User Attributes & Claims** form, click **Edit** > **Add new claim**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:109]:
```suggestion
4. Under the **Attributes & Claims** form, click **Edit** > **Add new claim**.
```

File: docs/semgrep-app/sso.md [hunk 11]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+    1. Enter `name` in the **Name** field.
+    2. From the **Source attribute** drop-down box, select `user.givenname`.
+    ![Screenshot of create your own Application page in Azure AD](/img/azure_ad-add-claim.png#bordered)
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:112]:
```suggestion
    ![Screenshot of create your own Application page in Azure AD](/img/azure_ad-add-claim.png#bordered)
    3. Click **Save**.
```

File: docs/semgrep-app/sso.md [hunk 12]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+4. Under the **User Attributes & Claims** form, click **Edit** < **Add new claim**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:113]:
```suggestion
4. Under the **Attributes & Claims** form, click **Edit** > **Add new claim**.
```

File: docs/semgrep-app/sso.md [hunk 13]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+    1. Enter `email` in the **Name** field.
+    2. From the **Source attribute** drop-down box, select `user.email`.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:115]:
```suggestion
    2. From the **Source attribute** drop-down box, select `user.email`.
    3. Click **Save**.
```

File: docs/semgrep-app/sso.md [hunk 14]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+6. Fill out the values required by Semgrep App's SAML2 form (**IdP SSO URL**, **IdP Issuer ID**, **X.509 Certificate**) by copying the values from Azure AD's **Set up Single Sign-On with SAML** page. Perform the following steps:
+    1. Copy the **Login URL** value from Azure AD to **IDP SSO URL**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:117]:
```suggestion
    1. Copy the **Login URL** value from Azure AD to **IDP SSO URL** in Semgrep App. 
```

File: docs/semgrep-app/sso.md [hunk 15]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+4. Under the **User Attributes & Claims** form, click **Edit** > **Add new claim**.
+    1. Enter `name` in the **Name** field.
+    2. From the **Source attribute** drop-down box, select `user.givenname`.
+    ![Screenshot of create your own Application page in Azure AD](/img/azure_ad-add-claim.png#bordered)
+4. Under the **User Attributes & Claims** form, click **Edit** < **Add new claim**.
+    1. Enter `email` in the **Name** field.
+    2. From the **Source attribute** drop-down box, select `user.email`.
+6. Fill out the values required by Semgrep App's SAML2 form (**IdP SSO URL**, **IdP Issuer ID**, **X.509 Certificate**) by copying the values from Azure AD's **Set up Single Sign-On with SAML** page. Perform the following steps:
+    1. Copy the **Login URL** value from Azure AD to **IDP SSO URL**. 
+    2. Copy **Azure AD Identifier** value to **IdP Issuer ID**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:118]:
```suggestion
    2. Copy **Azure AD Identifier** value to **IdP Issuer ID** in Semgrep App.
```

File: docs/semgrep-app/sso.md [hunk 16]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+    3. Click **Download** next to **Certificate (Base64)**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:119]:
```suggestion
    3. In **Set up Single Sign-On with SAML**, click **Download** next to **Certificate (Base64)**.
```

File: docs/semgrep-app/sso.md [hunk 17]
```diff
@@ -61,4 +61,70 @@ To set up SSO:
...
+    4. Copy the contents of the downloaded certificate to **X.509 Certificate**.
```
Comment (adamkvitek) [inline docs/semgrep-app/sso.md:120]:
```suggestion
    4. In Semgrep App, under **X.509 Certificate**, click **Browse** and then add the downloaded file. 
```

**Review by adamkvitek — APPROVED** [review]
@s-santillan @holdenmcg thank you for the amazing work here and all your patience and advice to test this procedure.

Contrary to what I said previously on call, this can be kept in the same file probably... if we need to prioritize the search feel free to move it and just include the link into `sso.md`.

I left small suggestions to clarify things where I was not sure, feel free to double check also and test this out again and rewrite them however you please. Thank you!

**Pull request merged**